cwl/src/main/scala/cwl/ExpressionEvaluator.scala

```diff
@@ -1,35 +1,53 @@
 package cwl
 
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.string.MatchesRegex
+import shapeless.Witness
 import wdl.util.JsUtil
 import wdl.values.WdlValue
 
 // http://www.commonwl.org/v1.0/CommandLineTool.html#Expressions
 object ExpressionEvaluator {
   // A code fragment wrapped in the $(...) syntax must be evaluated as a ECMAScript expression.
-  private val EcmaScriptRegex = """\$\((.*)\)""".r
+  val ECMAScriptExpressionWitness = Witness("""\$\((.*)\)""")
+  type MatchesECMAScript = MatchesRegex[ECMAScriptExpressionWitness.T]
+  type ECMAScriptExpression = String Refined MatchesRegex[ECMAScriptExpressionWitness.T]
 
   // A code fragment wrapped in the ${...} syntax must be evaluated as a ECMAScript function body for an anonymous,
   // zero-argument function.
-  private val EcmaFunctionRegex = """\$\{(.*)\}""".r
+  val ECMAScriptFunctionWitness = Witness("""\$\{(.*)\}""")
+  type ECMAScriptFunction = String Refined MatchesRegex[ECMAScriptFunctionWitness.T]
+  type MatchesECMAFunction = MatchesRegex[ECMAScriptFunctionWitness.T]
 
-  def evalExpression(expression: String, parameterContext: ParameterContext): WdlValue = {
-    val evalValues = Map(
-      "inputs" -> parameterContext.inputs,
-      "runtime" -> parameterContext.runtime,
-      "self" -> parameterContext.self
-    )
 
-    // TODO: WOM: Are we supposed to be trimming the expression just in case before matching?
-    expression.trim match {
-      case EcmaScriptRegex(ecmaScriptExpression) => JsUtil.eval(ecmaScriptExpression, evalValues)
-      case EcmaFunctionRegex(functionBody) =>
+  def evalExpression(expression: ECMAScriptExpression, parameterContext: ParameterContext): WdlValue = {
+    val ECMAScriptExpressionRegex = ECMAScriptExpressionWitness.value.r
+    expression.value match {
+      case ECMAScriptExpressionRegex(script) => JsUtil.eval(script, paramValues(parameterContext))
+      case _ => throw new RuntimeException(s"Expression was unable to be matched to Regex. This is never supposed to happen thanks to our JSON parsing library")
+    }
+  }
+
+  def evalFunction(function: ECMAScriptFunction, parameterContext: ParameterContext): WdlValue = {
+    val ECMAScriptFunctionRegex = ECMAScriptFunctionWitness.value.r
```

Comment:
Aha! I was wondering how to reuse a regex witness at compile time and runtime. Makes sense now. cc @mcovarr